From b78ec14242b7bf1c92bd5c1f81f30135811219af Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Sat, 2 Dec 2023 17:32:27 +0100
Subject: [PATCH 0804/1141] media: dcmipp: pixelproc: correct set_fmt handling

Correct pad number in v4l2_subdev_get_try_format calls and add
handling to update the src pad format when setting the sink pad format.

Change-Id: Ia5f0b87afab0ff3ac59387936133658e35f77f53
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/344612
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
---
 .../st/stm32/stm32-dcmipp/dcmipp-pixelproc.c  | 40 +++++++++++++------
 1 file changed, 27 insertions(+), 13 deletions(-)

--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelproc.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelproc.c
@@ -401,11 +401,11 @@ static int dcmipp_pixelproc_get_fmt(stru
 
 	if (IS_SINK(fmt->pad))
 		fmt->format = fmt->which == V4L2_SUBDEV_FORMAT_TRY ?
-			      *v4l2_subdev_get_try_format(sd, state, 0) :
+			      *v4l2_subdev_get_try_format(sd, state, fmt->pad) :
 			      pixelproc->sink_fmt;
 	else
 		fmt->format = fmt->which == V4L2_SUBDEV_FORMAT_TRY ?
-			      *v4l2_subdev_get_try_format(sd, state, 0) :
+			      *v4l2_subdev_get_try_format(sd, state, fmt->pad) :
 			      pixelproc->src_fmt;
 
 	mutex_unlock(&pixelproc->lock);
@@ -435,11 +435,35 @@ static int dcmipp_pixelproc_set_fmt(stru
 			pad_fmt = &pixelproc->src_fmt;
 
 	} else {
-		pad_fmt = v4l2_subdev_get_try_format(sd, state, 0);
+		pad_fmt = v4l2_subdev_get_try_format(sd, state, fmt->pad);
 	}
 
 	dcmipp_pixelproc_adjust_fmt(&fmt->format, fmt->pad);
 
+	/* When setting sink format, we have to update the src format */
+	if (IS_SINK(fmt->pad)) {
+		struct v4l2_mbus_framefmt *src_pad_fmt;
+
+		if (fmt->which == V4L2_SUBDEV_FORMAT_ACTIVE)
+			src_pad_fmt = &pixelproc->src_fmt;
+		else
+			src_pad_fmt = v4l2_subdev_get_try_format(sd, state, 1);
+
+		*src_pad_fmt = fmt->format;
+		if (fmt->format.code >= MEDIA_BUS_FMT_Y8_1X8 &&
+		    fmt->format.code < MEDIA_BUS_FMT_SBGGR8_1X8)
+			src_pad_fmt->code = MEDIA_BUS_FMT_YUYV8_2X8;
+		else
+			src_pad_fmt->code = MEDIA_BUS_FMT_RGB565_2X8_LE;
+
+		dev_dbg(pixelproc->dev, "%s: source format update: new:%dx%d (0x%x, %d, %d, %d, %d)\n",
+			pixelproc->sd.name,
+			src_pad_fmt->width, src_pad_fmt->height,
+			src_pad_fmt->code, src_pad_fmt->colorspace,
+			src_pad_fmt->quantization,
+			src_pad_fmt->xfer_func, src_pad_fmt->ycbcr_enc);
+	}
+
 	dev_dbg(pixelproc->dev, "%s: %s format update: old:%dx%d (0x%x, %d, %d, %d, %d) new:%dx%d (0x%x, %d, %d, %d, %d)\n",
 		pixelproc->sd.name,
 		PAD_STR(fmt->pad),
@@ -466,16 +490,6 @@ static int dcmipp_pixelproc_set_fmt(stru
 		pixelproc->compose.left = 0;
 		pixelproc->compose.width = fmt->format.width;
 		pixelproc->compose.height = fmt->format.height;
-
-		/* Update src pad format & size */
-		pixelproc->src_fmt = fmt_src_default;
-		if (fmt->format.code >= MEDIA_BUS_FMT_Y8_1X8 &&
-		    fmt->format.code < MEDIA_BUS_FMT_SBGGR8_1X8)
-			pixelproc->src_fmt.code = MEDIA_BUS_FMT_YUYV8_2X8;
-		else
-			pixelproc->src_fmt.code = MEDIA_BUS_FMT_RGB565_2X8_LE;
-		pixelproc->src_fmt.width = fmt->format.width;
-		pixelproc->src_fmt.height = fmt->format.height;
 	}
 
 out:
